File: leetcode/golang/src/keys-and-rooms.go
/* -------------------------------------------------------------------------- */
/*                https://leetcode.com/problems/keys-and-rooms                */
/* -------------------------------------------------------------------------- */

package main

import "fmt"

func CanVisitAllRooms(rooms [][]int) bool {
	n := len(rooms)

	visited := make([]bool, n)
	stack := []int{0}

	for len(stack) > 0 {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		fmt.Println(top)
		visited[top] = true
		for _, nbr := range rooms[top] {
			if !visited[nbr] {
				stack = append(stack, nbr)
			}
		}
	}

	for _, v := range visited {
		if !v {
			return false
		}
	}
	return true
}

File: leetcode/cpp/keys-and-rooms.cpp
/* -------------------------------------------------------------------------- */
/*                https://leetcode.com/problems/keys-and-rooms                */
/* -------------------------------------------------------------------------- */

#include<vector>
#include<stack>

using namespace std;

bool canVisitAllRooms(vector<vector<int>>& rooms) {
    int n = rooms.size();
    vector<bool> visited(n, false);

    stack<int> st;
    st.push(0);

    while (!st.empty()) {
        int top = st.top();
        st.pop();
        visited[top] = true;

        for (const auto& nbr : rooms[top]) {
            if (!visited[nbr]) {
                st.push(nbr);
            }
        }
    }

    for (const auto& each : visited) {
        if (!each) {
            return false;
        }
    }

    return true;
}